libs/server-sdk/src/data_sources/polling_data_source.cpp

```diff
@@ -0,0 +1,245 @@
+    boost::asio::any_io_executor const& ioc,
+    IDataSourceUpdateSink& handler,
+    DataSourceStatusManager& status_manager,
+    Logger const& logger)
+    : ioc_(ioc),
+      logger_(logger),
+      status_manager_(status_manager),
+      update_sink_(handler),
+      requester_(ioc),
+      timer_(ioc),
+      polling_interval_(
+          std::get<
+              config::shared::built::PollingConfig<config::shared::ServerSDK>>(
+              data_source_config.method)
+              .poll_interval),
+      request_(MakeRequest(data_source_config, endpoints, http_properties)) {
+    auto const& polling_config = std::get<
+        config::shared::built::PollingConfig<config::shared::ServerSDK>>(
+        data_source_config.method);
+    if (polling_interval_ < polling_config.min_polling_interval) {
+        LD_LOG(logger_, LogLevel::kWarn)
+            << "Polling interval too frequent, defaulting to "
+            << std::chrono::duration_cast<std::chrono::seconds>(
+                   polling_config.min_polling_interval)
+                   .count()
+            << " seconds";
+
+        polling_interval_ = polling_config.min_polling_interval;
+    }
+}
+
+void PollingDataSource::DoPoll() {
+    last_poll_start_ = std::chrono::system_clock::now();
+
+    auto weak_self = weak_from_this();
+    requester_.Request(request_, [weak_self](network::HttpResult const& res) {
+        if (auto self = weak_self.lock()) {
+            self->HandlePollResult(res);
+        }
+    });
+}
+
+void PollingDataSource::HandlePollResult(network::HttpResult const& res) {
+    auto header_etag = res.Headers().find("etag");
+    bool has_etag = header_etag != res.Headers().end();
+
+    if (etag_ && has_etag) {
+        if (etag_.value() == header_etag->second) {
+            // Got the same etag; we know the content has not changed.
+            // So we can just start the next timer.
+
+            // We don't need to update the "request_" because it would have
+            // the same Etag.
+            StartPollingTimer();
+            return;
+        }
+    }
+
+    if (has_etag) {
+        config::shared::builders::HttpPropertiesBuilder<
+            config::shared::ServerSDK>
+            builder(request_.Properties());
+        builder.Header("If-None-Match", header_etag->second);
+        request_ = network::HttpRequest(request_, builder.Build());
+
+        etag_ = header_etag->second;
+    }
+
+    if (res.IsError()) {
+        auto const& error_message = res.ErrorMessage();
+        status_manager_.SetState(
+            DataSourceStatus::DataSourceState::kInterrupted,
+            DataSourceStatus::ErrorInfo::ErrorKind::kNetworkError,
+            error_message.has_value() ? *error_message : "unknown error");
+        LD_LOG(logger_, LogLevel::kWarn)
+            << "Polling for feature flag updates failed: "
+            << (error_message.has_value() ? *error_message : "unknown error");
+    } else if (res.Status() == 200) {
```

Comment:
Kinda wish we could use the beast error codes here. Maybe we can, they might just be the numeric values of the status codes or have a getter for it.